tofhir-engine/src/main/scala/io/tofhir/engine/data/write/FileSystemWriter.scala

```diff
@@ -22,92 +26,156 @@ class FileSystemWriter(sinkSettings: FileSystemSinkSettings) extends BaseFhirWri
-            // | 2013-05-22 | NULL | [ {encounter-diagnosis} ] | {inactive, http...| {G40, Parkinson's disease} | Encounter/bb7134...| 63058b87a718e66d4...| [ {https://aiccele...| 2013-05-07 | Condition | {Patient/0b3a0b23... | NULL |
-            // +-----------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+------------+--------------------+--------------------+
-            val resourcesDF = spark.read.json(resourcesDS)
-
-            // Extend resourcesDF to handle partitioning if required.
-            // Some partition columns may not exist in the DataFrame (e.g., nested fields like `subject.reference`),
-            // so this extension adds the missing columns, allowing Spark to partition the DataFrame accordingly.
-            if(partitionColumns.isEmpty){
-              // If no partition columns are defined, return the DataFrame as-is
-              resourcesDF
+        // Check if the sink path is for HDFS
+        if (sinkSettings.path.startsWith("hdfs://")) {
+          // Extract the scheme (e.g., "hdfs") and authority (e.g., "namenode:9000") from the sinkSettings.path
+          val uri = new URI(sinkSettings.path)
+          val defaultFS = s"${uri.getScheme}://${uri.getAuthority}"
+
+          // Process each partition of the DataFrame
+          df.foreachPartition((partition: Iterator[FhirMappingResult]) => {
+            // Collect all elements in the current partition into a sequence
+            val fhirMappingResults: Seq[FhirMappingResult] = partition.toSeq
+            // Check if the partition contains any records
+            if (fhirMappingResults.nonEmpty) {
+              // Group the records by their resourceType
+              fhirMappingResults.groupBy(_.resourceType.get).foreach { case (resourceType, fhirResources) =>
+                // Log the number of resources and the resourceType being written
+                logger.debug("Will write {} {} resources to HDFS.", fhirResources.length, resourceType)
+                // Prepare the data to be written by concatenating resources into a single string, separated by newlines
+                val data = fhirResources.map(_.mappedFhirResource.get.mappedResource.get).mkString("\n")
+
+                // Initialize Hadoop Configuration
+                val conf = new Configuration()
+                conf.set("fs.defaultFS", defaultFS) // Set the HDFS default file system dynamically
+                // Get the HDFS FileSystem instance using the configuration
+                val fs = FileSystem.get(conf)
+
+                // Define the HDFS directory path for the current resourceType
+                val dirPath = new org.apache.hadoop.fs.Path(s"${sinkSettings.path}/$resourceType")
+                // Ensure the directory exists; create it if it does not
+                if (!fs.exists(dirPath)) {
+                  fs.mkdirs(dirPath)
```

Comment:
I suggest to keep an eye on the exception(s) that may be thrown if the file cannot be created on hadoop because of several reasons such as lack of permission.